vortex-io/src/read_ranges.rs

```diff
@@ -0,0 +1,150 @@
+#![allow(dead_code)]
+use std::future::Future;
+use std::io;
+use std::io::ErrorKind;
+use std::ops::Range;
+use std::sync::Arc;
+
+use bytes::Bytes;
+use futures_util::{stream, FutureExt, StreamExt, TryStreamExt};
+use vortex_error::VortexExpect;
+
+use crate::{Dispatch, IoDispatcher, VortexReadAt};
+
+const MAX_BUFFERED_READS: usize = 10;
+
+#[derive(Debug, Clone)]
+pub struct VortexReadRanges<R> {
+    read: R,
+    dispatcher: Arc<IoDispatcher>,
+    max_gap: usize,
+}
+
+impl<R> VortexReadRanges<R> {
+    pub fn new(read: R, dispatcher: Arc<IoDispatcher>, max_gap: usize) -> VortexReadRanges<R> {
+        Self {
+            read,
+            dispatcher,
+            max_gap,
+        }
+    }
+}
+
+impl<R: VortexReadAt> VortexReadRanges<R> {
+    pub fn read_byte_ranges(
+        &self,
+        ranges: Vec<Range<usize>>,
+    ) -> impl Future<Output = io::Result<Vec<Bytes>>> + 'static {
```

Comment:
We should probably update the signature to reflect that the returned future is `Send`, or perhaps just return a `BoxFuture` directly